File: node/hid-monitor-control.js
#!/usr/bin/env node

// export NODE_PATH=`npm root -g`

const { strict } = require('assert');
const util = require('util');

// const { HID } = require('node-hid');

const assert = require('assert').strict;

HID = require('node-hid');
HID.setDriverType('libusb');

device_id_pairs = [
    (0x056D, 0x4059) // EV2760
]

function get_input_source_table(model) {
    table = {
        'EV2760': {
            'DVI': 0x0200,
            'DisplayPort1': 0x0300,
            'DisplayPort2': 0x0301,
            'HDMI': 0x0400 }
    }
	return table[model]
}

function lookup_input_source_alias(input) {
	table = {
        'DP1': 'DisplayPort1',
        'DP2': 'DisplayPort2',
	}
	
	key = Object.keys(table).find(k => k.toLowerCase() == input.toLowerCase())
	if (key) {
		return table[key]
	}
	return input
}

function set_val(dev, num, val) {
	buf = [0x02, 0x01, 0xFF, num & 255, num >> 8, 0x00, 0x00, val & 255, val >> 8]
	buf = buf.concat(Array(39 - buf.length).fill(0))
	dev.sendFeatureReport(buf)

	tmp = dev.getFeatureReport(7, 8).slice(0, 8)
	assert.strictEqual(tmp[0], 7)
	assert.deepStrictEqual(tmp.slice(1,7), buf.slice(1, 7))
}

function get_val(dev, num) {
	buf = [ 0x03, 0x01, 0xFF, num & 255, num >> 8]
	buf = buf.concat(Array(39 - buf.length).fill(0))
	dev.sendFeatureReport(buf)

	tmp = dev.getFeatureReport(3, 40)
	assert.deepStrictEqual(tmp.slice(0, 7), buf.slice(0, 7))
	val = tmp[8] * 256 + tmp[7]

	tmp = dev.getFeatureReport(7, 8).slice(0, 8)
	assert.strictEqual(tmp[0], 7)
	assert.deepStrictEqual(tmp.slice(1,7), buf.slice(1, 7))
	return val
}

function print_usage(input_source_table=None) {
    console.log("Usage: python hid-monitor-control.py <INPUT> [INPUT2]")
    if (input_source_table) {
		console.log("       INPUT = " + Object.keys(input_source_table).join(' | '))
	}
    console.log("")
	console.log("       python hid-monitor-control.py switcher")
}

des = HID.devices().find(d => device_id_pairs.includes((d.vendorId, d.productId)))
// console.log(des)

hid = new HID.HID(des.vendorId, des.productId)

sn_pn = hid.getFeatureReport(8, 25);
sn_pn = sn_pn.map(c => String.fromCharCode(c)).join('')
serial_number = sn_pn.substring(1,9)
product_name = sn_pn.substring(9, 25).replace(/ +$/, '')
// console.log(serial_number, product_name)

// console.log('0x' + get_val(hid, 0x48).toString(16))

input_source_table = get_input_source_table(product_name)
// default to EV2760 because that is the only one known for now
if (!input_source_table) {
	input_source_table = get_input_source_table('EV2760')
}


if (process.argv.length > 4) {
	print_usage(input_source_table)
	process.exit(1)
}

if (process.argv.length == 2) {
	// console.log('0x' + get_val(hid, 0x48).toString(16))
	num = 1
	if (get_val(hid, 0x40) != 0) {
		num = 2
	}
	sels = Array(num).fill(undefined)
	for (var i = 0; i < num; i++) {
		if (num > 1) {
			set_val(hid, 0xF9, i)
		}
		val = get_val(hid, 0x48)
		val = Object.keys(input_source_table).find(k => input_source_table[k] == val)
		if (val instanceof Number) {
			val = "Unknown(" + val + ")"
		}
		sels[i] = val
	}
	console.log("Input: %s", sels.join(' '))
	print_usage(input_source_table)
	process.exit(0)
}

if (process.argv.length == 3 && process.argv[2].toLowerCase() != 'switcher') {
	sel = process.argv[2]
	sel = lookup_input_source_alias(sel)
	sel = Object.keys(input_source_table).find(k => k.toLowerCase() == sel.toLowerCase() )
	set_val(hid, 0x40, 0)
	set_val(hid, 0x48, input_source_table[sel])
	process.exit(0)
}

if (process.argv.length == 4) {
	num = process.argv.length - 2
	sels = process.argv.slice(2,4)
	for (var i = 0; i < num; i++) {
		sel = sels[i]
		sel = lookup_input_source_alias(sel)
		sel = Object.keys(input_source_table).find(k => k.toLowerCase() == sel.toLowerCase() )
		sels[i] = sel
	}
	set_val(hid, 0x40, 1)
	for (var i = 0; i < num; i++) {
		set_val(hid, 0xF9, i)
		set_val(hid, 0x48, input_source_table[sels[i]])
	}
	process.exit(0)
}

// switcher event loog
function data_handler(data) {
	console.log(data)
}

// hid.newListener('data', data_handler)

function data_handler(dat) {
	console.log(dat)
	if (dat.slice(0,5).equals(Buffer.of(0x03, 0x01, 0xFF, 0x3D, 0x00))) {
		sel = undefined
		btn = dat[8] * 256 + dat[7]
		switch (btn) {
			case 0x20: sel = 'DisplayPort1'; break
			case 0x10: sel = 'DisplayPort2'; break
			case 0x08: sel = 'HDMI'; break
			case 0x04: sel = 'DVI'; break
		}
		if (sel)  {
			set_val(hid, 0x48, input_source_table[sel])
		}
	}
}
hid.on('data', data_handler)

// hid.resume()

// node will keep running

File: ruby/hid-monitor-control.rb
#!ruby
require 'hidapi'
# https://rubygems.org/gems/hidapi/

device_id_pairs = [
  [0x056D, 0x4027], # EV2456
  [0x056D, 0x4014], # EV2750
  [0x056D, 0x4059], # EV2760
  [0x056D, 0x4065], # EV3895
]

$input_source_table = {


  :EV2456 => { :D-SUB => 0x0100, :DVI => 0x200, :DisplayPort => 0x0300, :HDMI => 0x0400 },
  :EV2750 => { :DVI => 0x0200, :DisplayPort => 0x0300, :HDMI => 0x0400 },
  :EV2760 => { :DVI => 0x0200, :DisplayPort1 => 0x0300, :DisplayPort2 => 0x0301, :HDMI => 0x0400 },
  :EV3895 => { :DisplayPort => 0x0300, :'USB-C' => 0x0301, :HDMI1 => 0x0400, :HDMI2 => 0x0401 },
}

$alias_table = {
  :DP => :DisplayPort,
  :DP1 => :DisplayPort1,
  :DP2 => :DisplayPort2,
  :TypeC => :'USB-C',
  :'Type-C' => :'USB-C',
  :USB => :'USB-C',
  :USBC => :'USB-C',
  :DSUB => :'D-SUB',
}

def print_usage(input_source_table = nil)
  STDERR.printf("\nUsage: %s INPUT1 [INPUT2]\n", $0)
  if input_source_table then
    STDERR.printf("       INPUT = ")
    STDERR.puts (input_source_table.collect{ |k,v|
      k.to_s
    }.join(" | "))
  end
  STDERR.printf("\n")
  STDERR.printf("       %s switcher\n", $0)
end

def get_input_val(input_source_table, arg)
  return nil if not arg
  arg = (($alias_table.find{|k,v|
    k.to_s.downcase == arg.downcase
  } || [])[1] || arg).to_s
  val = input_source_table.find{|k,v|
    k.to_s.downcase == arg.downcase
  }
  val = val[1] if val
  return val
end

def set_val(dev, num, val)
  buf = Array.new(39, 0)
  buf[0] = 0x02
  buf[1] = 0x01
  buf[2] = 0xFF
  buf[3] = num & 255
  buf[4] = num >> 8
  buf[5] = 0x00
  buf[6] = 0x00
  buf[7] = val & 255
  buf[8] = val >> 8
  dev.send_feature_report(buf)
  tmp = dev.get_feature_report(7, buffer_size = 8)
  raise if tmp[0].ord != 0x07
  6.times {|i| raise if tmp[i+1].ord != buf[i+1] }
end

def get_val(dev, num)
  buf = Array.new(39, 0)
  buf[0] = 0x03
  buf[1] = 0x01
  buf[2] = 0xFF
  buf[3] = num & 255
  buf[4] = num >> 8
  buf[5] = 0x00
  buf[6] = 0x00
  buf[7] = 0x00
  buf[8] = 0x00
  dev.send_feature_report(buf)
  tmp = dev.get_feature_report(3, buffer_size = 39)
  7.times {|i| raise if buf[i] != tmp[i].ord }
  val = tmp[8].ord * 256 + tmp[7].ord
  tmp = dev.get_feature_report(7, buffer_size = 8)
  raise if tmp[0].ord != 0x07
  6.times {|i| raise if tmp[i+1].ord != buf[i+1] }
  return val
end

# HIDAPI.engine.enumerate().each { |dev|
#   printf("%04X:%04X\n", dev.vendor_id, dev.product_id)
# }

dev = HIDAPI.engine.enumerate().find { |dev|
  device_id_pairs.include? [dev.vendor_id, dev.product_id]
}

exit 1 if not dev

# dev = HIDAPI.engine.get_device(0x056D, 0x4059)

dev.open
serial_number, model_number = (
  tmp = dev.get_feature_report(8, buffer_size = 25)
  [tmp[1..8], tmp[9...25].strip]
)

input_source_table = $input_source_table[model_number.to_sym]
input_source_table ||= $input_source_table[:EV2760] # default to EV2760

if not ARGV[0] then
  printf("%s (S/N: %s)\n", model_number, serial_number)
  if get_val(dev, 0x40) != 0
    set_val(dev, 0xF9, 0)
    val = get_val(dev, 0x48)
    printf("Input1: %s\n", input_source_table.invert[val])
    set_val(dev, 0xF9, 1)
    val = get_val(dev, 0x48)
    printf("Input2: %s\n", input_source_table.invert[val])
  else
    val = get_val(dev, 0x48)
    printf("Input: %s\n", input_source_table.invert[val])
  end

  STDOUT.flush()
  print_usage(input_source_table)

elsif ARGV[0].downcase == 'switcher'
  STDOUT.flush()
  loop {
    dat = dev.read()
    puts(dat.unpack('C*').collect{|i| format("0x%02X", i)}.join(",")); STDOUT.flush()
    if (dat[0..4] == [0x03, 0x01, 0xFF, 0x3D, 0x00].pack('C*')) then
      btn = dat[8].ord * 256 + dat[7].ord
      sel = nil
      if btn == 0x20 then
        sel = if input_source_table[:DisplayPort1] then :DisplayPort1 else :DisplayPort end
      elsif btn == 0x10 then
        sel = if input_source_table[:DisplayPort2] then :DisplayPort2 else nil end
      elsif btn == 0x08 then
        sel = :HDMI
      elsif btn == 0x04 then
        sel = :DVI
      end
      if sel then
        set_val(dev, 0x48, input_source_table[sel])
      end
    end
  }
else
  val = get_input_val(input_source_table, ARGV[0])
  val2 = get_input_val(input_source_table, ARGV[1])

  exit if not val

  if val2
    if val == val2
      STDOUT.flush()
      STDERR.printf("Error: Input1 and Input2 must be different.\n")
      exit 1
    end
    set_val(dev, 0x40, 1)
    set_val(dev, 0xF9, 0)
    set_val(dev, 0x48, val)
    set_val(dev, 0xF9, 1)
    set_val(dev, 0x48, val2)
  else
    set_val(dev, 0x40, 0)
    set_val(dev, 0x48, val)
  end

  dev.close
end
